### ISSUE TITLE
“ZeroDivisionError: float division by zero ！” When I train on my own dataset

### ISSUE THREAD

**Issue by 1005914341** [deepinsight/insightface#1448]
Training: 2021-03-31 11:28:15,352-Speed 415.63 samples/sec   Loss 36.7039   Epoch: 1   Global Step: 2000   Required: 1 hours
testing verification..
(472, 512)
infer time 1.2834219999999998
Traceback (most recent call last):
  File "train.py", line 132, in <module>
    main(args_)
  File "train.py", line 118, in main
    callback_verification(global_step, backbone)
  File "/root/arcface_torch/utils/utils_callbacks.py", line 48, in __call__
    self.ver_test(backbone, num_update)
  File "/root/arcface_torch/utils/utils_callbacks.py", line 27, in ver_test
    acc1, std1, acc2, std2, xnorm, embeddings_list = verification.test(
  File "/usr/local/lib/python3.8/dist-packages/torch/autograd/grad_mode.py", line 27, in decorate_context
    return func(*args, **kwargs)
  File "/root/arcface_torch/eval/verification.py", line 275, in test
    _, _, accuracy, val, val_std, far = evaluate(embeddings, issame_list, nrof_folds=nfolds)
  File "/root/arcface_torch/eval/verification.py", line 191, in evaluate
    val, val_std, far = calculate_val(thresholds,
  File "/root/arcface_torch/eval/verification.py", line 148, in calculate_val
    _, far_train[threshold_idx] = calculate_val_far(
  File "/root/arcface_torch/eval/verification.py", line 175, in calculate_val_far
    far = float(false_accept) / float(n_diff)
ZeroDivisionError: float division by zero
Killing subprocess 3993
Traceback (most recent call last):
  File "/usr/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/usr/local/lib/python3.8/dist-packages/torch/distributed/launch.py", line 340, in <module>
    main()
  File "/usr/local/lib/python3.8/dist-packages/torch/distributed/launch.py", line 326, in main
    sigkill_handler(signal.SIGTERM, None)  # not coming back
  File "/usr/local/lib/python3.8/dist-packages/torch/distributed/launch.py", line 301, in sigkill_handler
    raise subprocess.CalledProcessError(returncode=last_return_code, cmd=cmd)
subprocess.CalledProcessError: Command '['/usr/local/bin/python', '-u', 'train.py', '--local_rank=0']' returned non-zero exit status 1.

**Comment by jzfrank** [comment]
Hi, how did you prepare your verification dataset? Did you use the ones provided official website or prepared by yourself? 

https://github.com/deepinsight/insightface/issues/791#issuecomment-516375445

If you have prepared the dataset yourself, I guess this comment could help. 
TLDR: remember to use 10-fold cross-validation in generating pairs.txt, during the preparation of the validation dataset.

**Comment by linmeimei0512** [comment]
Hi @1005914341 
I have the same problem. Did you solve the problem? 
THX.

**Comment by 1005914341** [comment]
> Hi @1005914341
> I have the same problem. Did you solve the problem?
> THX.

这个一般是你的数据集问题，检查下是否有些数据缺少标签，或者是有些图片不存在，或者有问题。

**Comment by linmeimei0512** [comment]
Hi @1005914341 
I solve this problem. It's 10-fold cross-validation problem.
Thank you.

**Comment by nadaboulares** [comment]
@1005914341  how you solved this plz

**Comment by itsnine** [comment]
In case somebody has ZeroDivisionError in calculate_val_far function in verification.py, check your pairs file. It will be validated on 10-fold CV and every fold must contain pairs of both types(same and different).

**Comment by Fazyl03** [comment]
We can change code in verification.py to this:
if true_accept == 0 and n_same == 0:
        print("Here is an error of Float Division By Zero")
        val = 0
    else:
        val = float(true_accept) / float(n_same)
    
    if false_accept == 0 and n_diff == 0:
        print("Here is an error of Float Division By Zero")
        far = 0
    else:
        far = float(false_accept) / float(n_diff)
    return val, far
I didn't understand what that means but for me it worked.
